### PR TITLE
[ISSUE #3210] Enhanced nacos resttemplate response handler

### PR DESCRIPTION
**Pull request by Maijh97**
Please do not create a Pull Request without creating an issue first.

## What is the purpose of the change

For better scalability, abstract nacos RestTemplate response handler
#3210 

## Brief changelog

- Abstract `com.alibaba.nacos.common.http.client.ResponseHandler` to handle response conversion to corresponding `response` type

- Define the `AbstractResponseConverterHandler#convertResult` template method in `AbstractResponseConverterHandler` and let different converters implement it

## Verifying this change

XXXX

Follow this checklist to help us incorporate your contribution quickly and easily:

* [ ] Make sure there is a Github issue filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
* [ ] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
* [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
* [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/alibaba/nacos/tree/master/test).
* [ ] Run `mvn -B clean package apache-rat:check findbugs:findbugs -Dmaven.test.skip=true` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.

### CODE REVIEW

**Comment by Maijh97** [comment]
please help to review this pr, thanks @KomachiSion

**Comment by Maijh97** [comment]
I found that a test timed out and caused an error in CI,  Can you help to rerun the ci job again. @KomachiSion

File: common/src/main/java/com/alibaba/nacos/common/http/client/AbstractResponseConverterHandler.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import org.apache.http.HttpStatus;
+
+import java.lang.reflect.Type;
+
+/**
+ * Abstract response converter handler
+ *
+ * @author mai.jh
+ */
+public abstract class AbstractResponseConverterHandler<T> implements ResponseHandler<T> {
```
Comment (KomachiSion) [inline common/src/main/java/com/alibaba/nacos/common/http/client/AbstractResponseConverterHandler.java:31]: AbstractResponseHandler is enough, I think.

File: common/src/main/java/com/alibaba/nacos/common/http/client/BeanResponseConverter.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import java.io.InputStream;
+import java.lang.reflect.Type;
+
+/**
+ * bean response converter
+ * Mainly converter response type as bean type
+ *
+ * @author mai.jh
+ */
+public class BeanResponseConverter<T> extends AbstractResponseConverterHandler<T> {
```
Comment (KomachiSion) [inline common/src/main/java/com/alibaba/nacos/common/http/client/BeanResponseConverter.java:31]: BeanResponseHandler, I think

File: common/src/main/java/com/alibaba/nacos/common/http/client/HttpConverter.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+import java.util.Map;
+
+/**
+ * http converter
+ * For NacosRestTemplate and NacosAsyncRestTemplate,
+ * provide initialization and register of response converter
+ *
+ * @author mai.jh
+ */
+public class HttpConverter {
```
Comment (KomachiSion) [inline common/src/main/java/com/alibaba/nacos/common/http/client/HttpConverter.java:34]: abstract class `AbstractNacosRestTemplate`?

File: common/src/main/java/com/alibaba/nacos/common/http/client/HttpConverter.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    /**
+     * register customization Response Handler
+     * @param responseHandler {@link ResponseHandler}
+     */
+    public void registerResponseHandler(String responseHandlerType, ResponseHandler responseHandler) {
+        responseConverterMap.put(responseHandlerType, responseHandler);
+    }
+    
+    /**
+     * Select a response handler by responseType
```
Comment (KomachiSion) [inline common/src/main/java/com/alibaba/nacos/common/http/client/HttpConverter.java:54]: comment need an end '.', please  check all you comment.

File: common/src/main/java/com/alibaba/nacos/common/http/client/RestResultResponseConverter.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+import java.io.InputStream;
+import java.lang.reflect.Type;
+
+/**
+ * RestResult response converter
+ * Mainly converter response type as {@link RestResult} type.
+ *
+ * @author mai.jh
+ */
+public class RestResultResponseConverter<T> extends AbstractResponseConverterHandler<T> {
```
Comment (KomachiSion) [inline common/src/main/java/com/alibaba/nacos/common/http/client/RestResultResponseConverter.java:32]: RestResultResponseHandler.

File: common/src/main/java/com/alibaba/nacos/common/http/client/StringResponseConverter.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+import java.lang.reflect.Type;
+
+/**
+ * string response handler,
+ * Mainly converter response type as string type.
+ *
+ * @author mai.jh
+ */
+public class StringResponseConverter extends AbstractResponseConverterHandler<String> {
```
Comment (KomachiSion) [inline common/src/main/java/com/alibaba/nacos/common/http/client/StringResponseConverter.java:31]: StringResponseHandler

**Comment by Maijh97** [comment]
I have adjusted the code.please help to review this pr again, thanks @KomachiSion

File: common/src/main/java/com/alibaba/nacos/common/http/client/AbstractNacosRestTemplate.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+/**
+ * For NacosRestTemplate and NacosAsyncRestTemplate,
+ * provide initialization and register of response converter.
+ *
+ * @author mai.jh
+ */
+public abstract class AbstractNacosRestTemplate {
+    
+    private final Map<String, ResponseHandler> responseConverterMap = new HashMap<String, ResponseHandler>();
```
Comment (KomachiSion) [inline common/src/main/java/com/alibaba/nacos/common/http/client/AbstractNacosRestTemplate.java:35]: We need change the name, because We rename the class.

**Comment by KomachiSion** [comment]
We will merge this PR after 1.3.1 release

**Comment by KomachiSion** [comment]
@Maijh97 There are conflicting files. Please fix it.

**Comment by Maijh97** [comment]
> @Maijh97 There are conflicting files. Please fix it.

Code conflicts have been resolved, and some code style issues have been modified

**Pull request merged**